src/rgw/rgw_realm_reloader.cc

```diff
@@ -183,7 +183,10 @@ void RGWRealmReloader::reload()
    * the dynamic reconfiguration. */
   env.auth_registry = rgw::auth::StrategyRegistry::create(
       cct, implicit_tenants, env.driver);
-  env.lua.manager = env.driver->get_lua_manager();
+  env.lua.manager = env.driver->get_lua_manager(env.lua.manager->get_luarocks_path());
+  if (env.lua.background) {
+    env.lua.background->set_manager(env.lua.manager.get());
```

Comment:
even if this updates the `LuaManager` pointer, it leaves `Background::driver` as a dangling pointer to the Driver that we destroyed with `DriverManager::close_storage()` above. the `RadosLuaManager` would also hold a handling pointer to `RadosStore* const store`

we really should shut down this background thread before `close_storage()`, and only restart it once we've recreated the new driver and lua manager

Reply:
regarding the background thread. it is paused during reload (and the driver is set to null), and resumed only after the driver is updated to the new one. why do you think there is an issue?

regarding the lua package manager, it does not have any thread, just getting notifications from the rados thread (that should be paused during the reload). so, i'm not sure when the issue can happen?

Reply:
> why do you think there is an issue?

`env.lua.manager` extends the lifetime of `RadosLuaManager` past the point of its driver's deletion. this means that the `~RadosLuaManager()` dtor would be making calls on a `librados::IoCtx` whose `librados::Rados` client was already destroyed. we should free `env.lua.manager` before calling `close_storage()`